### PR TITLE
Add runtime representation of []v1.PreferredSchedulingTerm 

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

This object can be used for repeatedly scoring Nodes.
This allows an efficient implementation of #95738

**Special notes for your reviewer**:

Builds on top of #96064

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig scheduling

### CODE REVIEW

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go [hunk 1]
```diff
@@ -79,23 +79,11 @@ func (pl *NodeAffinity) Score(ctx context.Context, state *framework.CycleState,
...
-			// TODO: Avoid computing it for all nodes if this becomes a performance problem.
-			matches, err := corev1.MatchNodeSelectorTerms(node, &v1.NodeSelector{NodeSelectorTerms: []v1.NodeSelectorTerm{preferredSchedulingTerm.Preference}})
-			if err != nil {
-				return 0, framework.AsStatus(err)
-			}
-
-			if matches {
-				count += int64(preferredSchedulingTerm.Weight)
-			}
+		preferredNodeAffinity, err := nodeaffinity.NewPreferredSchedulingTerms(affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution)
```
Comment (damemi) [inline pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:82]: Does this still compute it for all nodes? Maybe I am missing where the computed affinity is stored, or that will be added in another PR, just curious

File: staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go [hunk 2]
```diff
@@ -113,12 +100,83 @@ func (ns *LazyErrorNodeSelector) Match(node *v1.Node) (bool, error) {
...
+	terms []preferredSchedulingTerm
+}
+
+// NewPreferredSchedulingTerms returns a PreferredSchedulingTerms or all the parsing errors found.
+// If a v1.PreferredSchedulingTerm has a 0 weight, its parsing is skipped.
+func NewPreferredSchedulingTerms(terms []v1.PreferredSchedulingTerm) (*PreferredSchedulingTerms, error) {
+	var errs []error
+	parsedTerms := make([]preferredSchedulingTerm, 0, len(terms))
+	for _, term := range terms {
+		if term.Weight == 0 || isEmptyNodeSelectorTerm(&term.Preference) {
```
Comment (damemi) [inline staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go:114]: could the `term.Weight == 0` check just be added to `isEmptyNodeSelectorTerm`?

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:82]: Correct. For that, I would need to implement PreFilter, which is out of the scope for this PR. But I might leave that for 1.21. The priority is implementing #95738, for which parsing happens during plugin instantiation.

File: staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go (hunk 2 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go:114]: Not really. Not that isEmptyNodeSelectorTerm receives v1.NodeSelectorTerm and is also used in the hard node affinity representation.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go (hunk 1 above)
Comment (damemi) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:82]: Sounds good, just wanted to make sure I understood thanks

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:83]: I created an issue and restored the TODO
